agent/lib_monolog.c

```diff
@@ -0,0 +1,387 @@
+/*
+ * Copyright 2022 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "php_agent.h"
+#include "php_call.h"
+#include "php_user_instrument.h"
+#include "php_datastore.h"
+#include "php_execute.h"
+#include "php_hash.h"
+#include "php_wrapper.h"
+#include "fw_hooks.h"
+#include "fw_support.h"
+#include "nr_datastore_instance.h"
+#include "nr_segment_datastore.h"
+#include "util_logging.h"
+#include "util_memory.h"
+#include "util_strings.h"
+#include "util_sleep.h"
+
+/*
+ * Purpose : Return a copy of Monolog\Logger::API.
+ *
+ * Params  : 1. An instance of Monolog\Logger.
+ *
+ * Returns : The Monolog version number as an integer
+ */
+static int nr_monolog_version(zval* logger TSRMLS_DC) {
+  int retval = 0;
+  zval* api = NULL;
+  zend_class_entry* ce = NULL;
+
+  if (0 == nr_php_is_zval_valid_object(logger)) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: Logger object is invalid", __func__);
+    return 0;
+  }
+
+  ce = Z_OBJCE_P(logger);
+  if (NULL == ce) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: Logger has NULL class entry",
+                     __func__);
+    return 0;
+  }
+
+  api = nr_php_get_class_constant(ce, "API");
+  if (NULL == api) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: Logger does not have API", __func__);
+    return 0;
+  }
+
+  if (nr_php_is_zval_valid_integer(api)) {
+    retval = Z_LVAL_P(api);
+  } else {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: expected API be a valid int, got type %d", __func__,
+                     Z_TYPE_P(api));
+  }
+
+  nr_php_zval_free(&api);
+  return retval;
+}
+
+/*
+ * Purpose : Convert $level argument of Monolog\Logger::addRecord to a string
+ * representation of Monolog's log level.
+ *
+ * Params  : Logger instance and Monolog\Logger::addRecord argument list
+ *
+ * Returns : A new string with Monolog's log level name
```

Comment:
Is the caller responsible for freeing?

Reply:
Yes.

Reply:
@zsistla This has been addressed in 7cbdff18a4ac99dc5e165d7c1f000140f1b851b6